streams-tests/jvm/src/test/scala/zio/stream/ZStreamPlatformSpecificSpec.scala

```diff
@@ -294,56 +295,59 @@ object ZStreamPlatformSpecificSpec extends ZIOBaseSpec {
-
-            receive <- refOut.get.repeatWhileZIO(s => ZIO.succeed(s.isEmpty))
-
-            _ <- server.interrupt
-          } yield assert(receive)(equalTo(message))
-        }),
-        test("write data")(check(Gen.string.filter(_.nonEmpty)) { message =>
-          (for {
-            refOut <- Ref.make("")
-
-            server <- ZStream
-                        .fromSocketServer(8897)
-                        .foreach(c => ZStream.fromIterable(message.getBytes).run(c.write))
-                        .fork
-
-            _ <- socketClient(8897).use { c =>
-                   val buffer = ByteBuffer.allocate(message.getBytes.length)
-
-                   ZIO
-                     .fromFutureJava(c.read(buffer))
-                     .repeatUntil(_ < 1)
-                     .flatMap { _ =>
-                       (buffer: Buffer).flip()
-                       refOut.update(_ => new String(buffer.array))
-                     }
-                 }.retry(Schedule.forever)
+                              }
+                              .take(messages.size.toLong)
+                              .runCollect
+          } yield assert(readMessages)(hasSameElementsDistinct(readMessages))
+        },
+        test("write data") {
+          for {
+            messages <- Gen.string1(Gen.unicodeChar).runCollectN(200)
+            writtenMessages <- ZStream
+                                 .fromSocketServer(8897)
+                                 .zip(ZStream.managed(socketClient(8897)))
+                                 .flatMap { case (serverChannel, clientChannel) =>
+                                   ZStream
+                                     .fromIterable(messages)
+                                     .mapZIO(m =>
+                                       ZStream.fromIterable(m.getBytes("UTF-8")).run(serverChannel.write) *> {
+                                         val buffer = ByteBuffer.allocate(m.getBytes("UTF-8").length)
 
-            receive <- refOut.get.repeatWhileZIO(s => ZIO.succeed(s.isEmpty))
+                                         ZIO
+                                           .fromFutureJava(clientChannel.read(buffer))
+                                           .repeatUntil(_ < 1)
+                                           .map { _ =>
+                                             (buffer: Buffer).flip()
+                                             new String(buffer.array)
+                                           }
+                                       }
+                                     )
+                                 }
+                                 .take(messages.size.toLong)
+                                 .runCollect
+          } yield assert(writtenMessages)(hasSameElementsDistinct(messages))
 
-            _ <- server.interrupt
-          } yield assert(receive)(equalTo(message)))
-        })
-      ),
+        }
+      ) @@ flaky(20), // socket connections can be flaky some times
```

Comment:
only needed for CI, was not able to make it fail locally - tried both my Linux box on Intel and MacOS on Apple Silicon

Reply:
Same experience here. Getting this merged will make it easier to complete some other Spec refactoring.